src/System.Management.Automation/engine/runtime/Operations/MiscOps.cs

```diff
@@ -311,6 +315,89 @@ internal static class PipelineOps
+                    paramBlock: null,
+                    beginBlock: null,
+                    processBlock: (NamedBlockAst)sbAst.EndBlock.Copy(),
+                    endBlock: null,
+                    dynamicParamBlock: null);
+                newScriptBlockAst.PostParseChecksPerformed = sbAst.PostParseChecksPerformed;
+                sbAst.Parent.SetParent(newScriptBlockAst);
+
+                return new ScriptBlock(newScriptBlockAst, isFilter: false);
+            };
+
+        private static bool TryRewriteForEachObjectCommand(
+            ExecutionContext context,
+            SessionStateInternal commandSessionState,
+            CommandParameterInternal[] commandElements,
+            ref CommandProcessorBase commandProcessor,
+            ref int commandIndex)
+        {
+            const string ForEachObject_ProcessParam = "Process";
+
+            if (commandProcessor.CommandInfo is CmdletInfo cmdlet && cmdlet.ImplementingType == typeof(ForEachObjectCommand))
+            {
+                int indexAdvanceOffset = 0;
+                int cmdElementsLength = commandElements.Length;
+                ScriptBlock processScriptBlock = null;
+
+                if (commandIndex == cmdElementsLength - 1)
+                {
+                    var currentElement = commandElements[commandIndex];
+                    if (currentElement.ArgumentSpecified && !currentElement.ArgumentSplatted &&
+                        (!currentElement.ParameterAndArgumentSpecified || string.Equals(currentElement.ParameterName, "Process", StringComparison.OrdinalIgnoreCase)))
+                    {
+                        processScriptBlock = currentElement.ArgumentValue as ScriptBlock;
+                        indexAdvanceOffset = 1;
+                    }
+                }
+                else if (commandIndex == cmdElementsLength - 2)
+                {
+                    var currentElement = commandElements[commandIndex];
+                    var nextElement = commandElements[commandIndex + 1];
+
+                    if (currentElement.ParameterNameSpecified && !currentElement.ArgumentSpecified &&
+                        string.Equals(currentElement.ParameterName, ForEachObject_ProcessParam, StringComparison.OrdinalIgnoreCase) &&
+                        nextElement.ArgumentSpecified && !nextElement.ArgumentSplatted && !nextElement.ParameterNameSpecified)
+                    {
+                        processScriptBlock = nextElement.ArgumentValue as ScriptBlock;
+                        indexAdvanceOffset = 2;
+                    }
+                }
+
+                if (processScriptBlock != null && processScriptBlock.Ast is ScriptBlockAst sbAst)
+                {
+                    if (!sbAst.IsConfiguration && sbAst.ParamBlock == null && sbAst.BeginBlock == null &&
+                        sbAst.ProcessBlock == null && sbAst.DynamicParamBlock == null && sbAst.EndBlock != null &&
+                        sbAst.EndBlock.Unnamed)
```

Comment:
Do you consider moving this in parse time? Maybe we could have a special ast type? I am thinking about Where-Object where we could use the same way to optimize performance.

Reply:
> it would require adding new elements to the existing script block, and that will cause issue to the accuracy of extent, debugging, and more, which would be too much magic to deal with.

I see currently we skip optimization if we debug so it is not problem. 
If we look https://learn-powershell.net/2015/02/20/a-look-at-filtering-performance-in-powershell/ we can see that filter function has best performance. We could convert scriptblock to a filter function. This conversion does not look much magic. If we could still add parameter binding optimization we could get great performance win for most of common scenarios.

Reply:
Perhaps it is more easy at parse time as I already mentioned, I did something like while did experiments with class getters/setters. (We could enhance Parser/Compiler to support this and it could become a common approach for dynamic optimizations. It might be similar to what CoreCLR does today by postponing compilation and optimizations.)

Ah, I found that we have #9941 tracking issue for Where-Object.